cmds/legacy_boot/legacy_boot.go

```diff
@@ -0,0 +1,395 @@
+// Copyright 2012-2017 the u-root Authors. All rights reserved
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+// legacy_boot allows to handover a system running linuxboot/u-root
+// to a legacy preinstalled operating system by replacing the traditional
+// bootloader path
+
+//
+// Synopsis:
+//	legacy_boot
+//
+// Description:
+//	If returns to u-root shell, the code didn't found a local bootable option
+//
+// Notes:
+//	The code is looking for boot/grub/grub.cfg file as to identify the
+//	boot option.
+//	The first bootable device found in the block device tree is the one used
+//	Windows is not supported (that is a work in progress)
+//
+// Example:
+//	legacy_boot -v 	- Start the script in verbose mode for debugging purpose
+
+package main
+
+import (
+	"bufio"
+	"fmt"
+	"github.com/u-root/u-root/pkg/kexec"
+	"io"
+	"io/ioutil"
+	"log"
+	"os"
+	"strings"
+	"flag"
+	"syscall"
+)
+
+var verbose bool
+type options struct {
+        verbose bool
+}
+
+func check(e error) {
+	if e != nil {
+		panic(e)
+	}
+}
+func check_fs(e error) {
+}
+
+func blkDevicesList(blkpath string, devpath string) []string {
+	var blkDevices []string
+	files, err := ioutil.ReadDir(blkpath)
+	if err != nil {
+		log.Fatal(err)
+	}
+	for _, file := range files {
+		deviceEntry, err := ioutil.ReadDir(blkpath + file.Name() + devpath)
+		if err != nil {
+			if ( verbose ) {
+				println("can t read directory")
+			}
+			continue
+		}
+		blkDevices = append(blkDevices, deviceEntry[0].Name())
+	}
+	return (blkDevices)
+}
+
+// checkForBootableMbr is looking for bootable MBR signature 
+// Current support is limited to Hard disk devices and USB devices
+func checkForBootableMbr(path string) int {
+	var b511, b510 byte
+	var err error
+	f, err := os.Open(path)
+	check(err)
+	b1 := make([]byte, 512)
+	_,err=f.Read(b1)
+	check(err)
+	b511 = b1[511]
+	b510 = b1[510]
+	err=f.Close()
+	check(err)
+	if ((b511 == 0xaa) && (b510 == 0x55)) == true {
+		return 1
+	}
+	return 0
```

Comment:
my suggestion:


diff --git a/cmds/legacy_boot/legacy_boot.go b/cmds/legacy_boot/legacy_boot.go
index 549b4d5..6b578cd 100644
--- a/cmds/legacy_boot/legacy_boot.go
+++ b/cmds/legacy_boot/legacy_boot.go
@@ -26,17 +26,24 @@ package main
 
 import (
 	"bufio"
+	"encoding/binary"
 	"fmt"
 	"github.com/u-root/u-root/pkg/kexec"
 	"io"
 	"io/ioutil"
 	"log"
 	"os"
+	"path/filepath"
 	"strings"
 	"flag"
 	"syscall"
 )
 
+const (
+	bootableMBR = 0xaa55
+	signatureOffset = 510
+)
+
 var verbose bool
 type options struct {
         verbose bool
@@ -69,24 +76,22 @@ func blkDevicesList(blkpath string, devpath string) []string {
 	return (blkDevices)
 }
 
-// checkForBootableMbr is looking for bootable MBR signature 
+// checkForBootableMBR is looking for bootable MBR signature 
 // Current support is limited to Hard disk devices and USB devices
-func checkForBootableMbr(path string) int {
-	var b511, b510 byte
-	var err error
+func checkForBootableMBR(path string) (bool, error) {
+	var sig uint16
 	f, err := os.Open(path)
-	check(err)
-	b1 := make([]byte, 512)
-	_,err=f.Read(b1)
-	check(err)
-	b511 = b1[511]
-	b510 = b1[510]
-	err=f.Close()
-	check(err)
-	if ((b511 == 0xaa) && (b510 == 0x55)) == true {
-		return 1
+	if err != nil {
+		return false, err
+	}
+
+	if err := binary.Read(io.NewSectionReader(f, signatureOffset, 2), binary.LittleEndian, &sig); err != nil {
+		return false, err
 	}
-	return 0
+	if sig != bootableMBR {
+		return false, nil
+	}
+	return true, nil
 }
 
 // getDevicePartList returns all devices attached to a specific name like /dev/sdaX where X can move from 0 to 127
@@ -368,7 +373,13 @@ func main() {
 	// devices which do have such support
 	// drive are easy to detect
 	for _, entry := range blkList {
-		if checkForBootableMbr("/dev/"+entry) == 1 {
+		dev := filepath.Join("/dev", entry)
+		b, err := checkForBootableMBR(dev)
+		if err != nil {
+			// Not sure it matters; there can be many bogus entries?
+			log.Printf("MBR for %s failed: %v", dev, err)
+		}
+		if b {
 			fmt.Println("Bootable device found")
 			// We need to loop on the device entries which are into /dev/<device>X
 			// and mount each partitions as to find /boot entry if it is available somewhere

Reply:
oh, interesting to see what github did to my paste. I'll send it offline.